<launch>
    <arg name="odometry_topic"/>
    <arg name="camera_pose_topic"/>
    <arg name="depth_topic"/>
    <arg name="cloud_topic"/>
    <arg name="cx"/>
    <arg name="cy"/>
    <arg name="fx"/>
    <arg name="fy"/>
    <arg name="max_vel"/>
    <arg name="max_acc"/>
    <arg name="max_jer"/>
    <arg name="max_sna"/>
    <!-- <arg name="planning_horizon"/> -->
    <arg name="flight_type"/>
    <arg name="use_multitopology_trajs"/>
    <arg name="drone_id"/>
    <!-- main node -->
    <!-- <node pkg="ego_planner" name="drone_$(arg drone_id)_ego_planner_node" type="ego_planner_node" output="screen" launch-prefix="valgrind"> -->
    <node pkg="mission_fsm" name="drone_$(arg drone_id)_ego_planner_node" type="mission_fsm_node" output="screen">
        <remap from="~odom_world" to="/drone_$(arg drone_id)_$(arg odometry_topic)"/>
        <remap from="~mandatory_stop" to="/mandatory_stop_to_planner"/>
        <remap from="~planning/trajectory" to = "/drone_$(arg drone_id)_planning/trajectory"/>
        <remap from="~planning/data_display" to = "/drone_$(arg drone_id)_planning/data_display"/>
        <remap from="~planning/yaw_cmd" to = "/drone_$(arg drone_id)_planning/yaw_cmd"/>
        <remap from="~planning/broadcast_traj_send" to = "/broadcast_traj_from_planner"/>
        <remap from="~planning/broadcast_traj_recv" to = "/bridge/broadcast_traj_from_planner"/>
        <remap from="~planning/heartbeat" to = "/drone_$(arg drone_id)_traj_server/heartbeat"/>
        <!-- <remap from="/goal" to = "/goal_with_id"/> -->
        <!-- <remap from="/goal" to = "/move_base_simple/goal"/> -->
        <remap from ="/trigger" to="/move_base_simple/goal"/>
        <remap from="~grid_map/odom" to="/drone_$(arg drone_id)_$(arg odometry_topic)"/>
        <remap from="~grid_map/cloud" to="/mid360/points"/>
        <!-- <remap from="~grid_map/cloud" to="/livox/lidar"/> -->
        <remap from="~grid_map/pose"   to = "/drone_$(arg drone_id)_$(arg camera_pose_topic)"/>
        <remap from="~grid_map/depth" to = "/drone_$(arg drone_id)_$(arg depth_topic)"/>
        <remap from="/position_cmd" to="drone_$(arg drone_id)_planning/pos_cmd"/>
        <remap from="/tracking_target" to="/target/odom_visualization/pose"/>
        <!-- traj_server -->
        <param name="traj_server/time_forward" value="1.0" type="double"/>
        <!-- planning fsm -->
        <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
        <param name="fsm/ground_height_measurement" value="true" type="bool"/>
        <!-- <param name="fsm/thresh_replan_time" value="1.0" type="double"/> -->
        <!-- <param name="fsm/planning_horizon" value="$(arg planning_horizon)" type="double"/> -->
        <!--always set to 1.5 times grater than sensing horizen-->
        <param name="fsm/drone_id" value="$(arg drone_id)"/>
        <param name="fsm/arrive_dis_thr" value="0.2"/>
        <param name="fsm/odom_buffer_dis_threshold" value="0.05"/>
        <param name="fsm/follower_dis" value="0.8"/>
        <param name="fsm/emergency_time" value="1.0" type="double"/>
        <param name="fsm/realworld_experiment" value="false"/>
        <param name="fsm/fail_safe" value="true"/>
        <param name="fsm/battery_thr" value="19.0" type="double"/>

        <!-- <param name="grid_map/resolution" value="0.1" /> -->
        <!-- local_map的size必须比depth_filter_maxdist大两个resoluiton -->
        <param name="grid_map/local_update_range_x" value="8.0" />
        <param name="grid_map/local_update_range_y" value="8.0" />
        <param name="grid_map/local_update_range_z" value="4.0" />
        <!-- <param name="grid_map/obstacles_inflation" value="0.5" /> -->
        <param name="grid_map/obstacles_inflation" value="0.2" />

        <param name="grid_map/local_map_margin" value="10"/>
        <param name="grid_map/enable_virtual_wall" value="true"/>
        <param name="grid_map/virtual_ceil" value="2.0"/>
        <param name="grid_map/virtual_ground" value="-0.1"/>
        <param name="grid_map/visual_inflat_map_height" value="2.0"/>

        <!-- camera parameter -->
        <param name="grid_map/cx" value="$(arg cx)"/>
        <param name="grid_map/cy" value="$(arg cy)"/>
        <param name="grid_map/fx" value="$(arg fx)"/>
        <param name="grid_map/fy" value="$(arg fy)"/>
        <!-- <param name="grid_map/k_depth_scaling_factor" value="1000.0"/> -->
        <param name="grid_map/k_depth_scaling_factor" value="1.0"/>
        <param name="grid_map/skip_pixel" value="2"/>
        <param name="grid_map/use_depth_filter" value="true"/>
        <param name="grid_map/depth_filter_maxdist"   value="6.0"/>
        <param name="grid_map/depth_filter_mindist"   value="0.2"/>
        <param name="grid_map/depth_filter_margin"    value="2"/>
        <!-- local fusion -->
        <!-- <param name="grid_map/p_hit" value="0.90"/>
        <param name="grid_map/p_miss" value="0.45"/> -->
        <param name="grid_map/p_hit" value="0.90"/>
        <param name="grid_map/p_miss" value="0.45"/>
        <param name="grid_map/p_min" value="0.12"/>
        <param name="grid_map/p_max" value="0.90"/>
        <param name="grid_map/p_occ" value="0.80"/>
        <param name="grid_map/fading_time" value="-1.0"/>
        <param name="grid_map/min_ray_length" value="0.1"/>
        <!-- <param name="grid_map/max_ray_length" value="10.0"/> -->
        <param name="grid_map/max_ray_length" value="6.0"/>
        <param name="grid_map/show_occ_time" value="false"/>
        <param name="grid_map/pose_type" value="1"/>
        <param name="grid_map/frame_id" value="world"/>
        <param name="grid_map/hit_accum_max" value="40" />
        <param name="grid_map/hit_accum_threshold" value="10" />

        <!-- Perception utils -->
        <param name="perception_utils/top_angle" value="0.6" type="double"/>;
        <param name="perception_utils/left_angle" value="0.76" type="double"/>;
        <param name="perception_utils/right_angle" value="0.76" type="double"/>;
        <param name="perception_utils/max_dist" value="6.0" type="double"/>;
        <param name="perception_utils/vis_dist" value="1.0" type="double"/>;

        <!-- rc trcl params -->
        <param name="rc_ctrl/odom_topic" value="/drone_0_visual_slam/odom" type="string"/>
        <param name="rc_ctrl/rc_topic" value="/mavros/rc/in" type="string"/>
        <param name="rc_ctrl/px4ctrl_cmd_topic" value="/position_cmd" type="string"/>
        <param name="rc_ctrl/loop_rate" value="60.0" type="double"/>
        <param name="rc_ctrl/max_vel" value="1.0" type="double"/>
        <param name="rc_ctrl/max_acc" value="1.0" type="double"/>
        <param name="rc_ctrl/max_yaw_rate" value="0.8" type="double"/>
        <param name="rc_ctrl/rc_timeout" value="0.5" type="double"/>
        <param name="rc_ctrl/odom_timeout" value="0.1" type="double"/>
        <param name="rc_ctrl/rc_reverse_pitch" value="true" type="bool"/>
        <param name="rc_ctrl/rc_reverse_roll" value="false" type="bool"/>
        <param name="rc_ctrl/rc_reverse_yaw" value="true" type="bool"/>
        <param name="rc_ctrl/rc_reverse_throttle" value="false" type="bool"/>


        <!-- Fsm -->
        <!-- if time to the end of the traj < thresh_replan1 -->
        <param name="fsm/thresh_replan1" value="0.5" type="double"/>
        <!-- if aim Frontier Covered & time > thresh_replan2 -->
        <param name="fsm/thresh_replan2" value="1.5" type="double"/>
        <!-- if time > thresh_replan3 -->
        <param name="fsm/thresh_replan3" value="10.0" type="double"/>

        <param name="fsm/replan_time" value="0.200" type="double"/>

        <!-- planner manager -->
        <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
        <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
        <!-- <param name="manager/polyTraj_piece_length" value="1.25" type="double"/> -->
        <!-- <param name="manager/planning_horizon" value="$(arg planning_horizon)" type="double"/> -->
        <param name="manager/use_multitopology_trajs" value="$(arg use_multitopology_trajs)" type="bool"/>
        <param name="manager/drone_id" value="$(arg drone_id)"/>
        <!-- trajectory optimization -->
        <!-- <param name="optimization/constraint_points_perPiece" value="5" type="int"/> -->
        <param name="optimization/weight_obstacle" value="10000.0" type="double"/>
        <param name="optimization/weight_obstacle_soft" value="0.0" type="double"/>
        <param name="optimization/weight_trust_region" value="10.0" type="double"/>
        <param name="optimization/weight_curve_fitting" value="1000.0" type="double"/>
        <param name="optimization/weight_plane" value="10.0" type="double"/>
        <param name="optimization/weight_swarm" value="10000.0" type="double"/>
        <param name="optimization/weight_feasibility" value="250.0" type="double"/>
        <param name="optimization/weight_sqrvariance" value="0.0" type="double"/>
        <param name="optimization/weight_time" value="10.0" type="double"/>
        <param name="optimization/obstacle_clearance" value="0.1" type="double"/>
        <param name="optimization/obstacle_clearance_soft" value="1.5" type="double"/>
        <param name="optimization/swarm_clearance" value="0.45" type="double"/> <!-- heterogeneous support: my required clearance -->
        <!-- <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
        <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/> -->
        <param name="optimization/max_jer" value="$(arg max_jer)" type="double"/>
        <param name="optimization/max_sna" value="$(arg max_sna)" type="double"/>
    </node>
</launch>
